File: js/t7/tests/model.unit.test.js
import { describe, test } from 'node:test';
import assert from 'node:assert/strict';
import { T } from '../model.js';
import * as TI from './testInstance.js';
import { oracle7 } from './oracle.js';
import { assertSnapshotsEqual } from '../../t1/tests/oracle.js';

function makeEngine() {
  return T(TI.Piece, TI.InitialMainGrid, TI.ForbiddenGrid, TI.RotGrid, TI.InitialY, TI.InitialX,
           TI.PW, TI.FY, TI.FX, TI.NextLen, TI.Player, TI.HostIndex);
}

function bagsFnFrom(bag) { return () => bag; }

function makeHolesFn(w) {
  const holes = [];
  return (y) => {
    while (holes.length <= y) holes.push(Math.floor(Math.random() * w));
    const x = holes[y];
    if (!(0 <= x && x < w)) throw new Error(`makeHolesFn: hole out of range: ${x}`);
    return x;
  };
}

describe('Machine — field shape', () => {
  test('has garbage/target/gameoverView/connectedView; has no connected/messages field (§6.9)', () => {
    const eng = makeEngine();
    const m = new eng.Machine(0, bagsFnFrom(['A', 'B']));
    assert.strictEqual(typeof m.garbage, 'number');
    assert.strictEqual(typeof m.target, 'number');
    assert.ok(Array.isArray(m.gameoverView));
    assert.ok(Array.isArray(m.connectedView));
    assert.strictEqual(m.gameoverView.length, TI.Player.length);
    assert.strictEqual(m.connectedView.length, TI.Player.length);
    assert.ok(!('connected' in m), 'connected must not be a Machine field (§6.9)');
    assert.ok(!('messages' in m), 'messages must not be a Machine field (§6.9)');
  });

  test('s6 is nested, not flattened (unlike T2–T6\'s own inner engines)', () => {
    const eng = makeEngine();
    const m = new eng.Machine(0, bagsFnFrom(['A', 'B']));
    assert.ok('s6' in m);
    assert.ok(!('mg' in m) || typeof Object.getOwnPropertyDescriptor(m, 'mg') === 'undefined',
      'mg is a read-through getter, not a flattened own field');
    assert.deepStrictEqual(m.mg, m.s6.s1.mg);
  });

  test('checkAxioms rejects PlayerCount <= 1 and an out-of-range HostIndex', () => {
    const eng1 = T(TI.Piece, TI.InitialMainGrid, TI.ForbiddenGrid, TI.RotGrid, TI.InitialY, TI.InitialX,
                   TI.PW, TI.FY, TI.FX, TI.NextLen, [0], 0);
    const originalAssert = console.assert;
    let tripped = false;
    console.assert = (cond) => { if (!cond) tripped = true; };
    try {
      eng1.checkAxioms();
      assert.strictEqual(tripped, true, 'PlayerCount = 1 must trip an axiom');

      tripped = false;
      const eng2 = T(TI.Piece, TI.InitialMainGrid, TI.ForbiddenGrid, TI.RotGrid, TI.InitialY, TI.InitialX,
                     TI.PW, TI.FY, TI.FX, TI.NextLen, TI.Player, 99);
      eng2.checkAxioms();
      assert.strictEqual(tripped, true, 'out-of-range HostIndex must trip an axiom');
    } finally {
      console.assert = originalAssert;
    }
  });
});

describe('GeneratedGarbage / GenRemGarbage — exact arithmetic (req-multi-garbage-gen/cancel)', () => {
  test('normal garbage: c - 1 for c < 4, floored at 0; c itself for c >= 4', () => {
    const eng = makeEngine();
    assert.strictEqual(eng.generatedGarbage(0, false), 0);
    assert.strictEqual(eng.generatedGarbage(1, false), 0);
    assert.strictEqual(eng.generatedGarbage(2, false), 1);
    assert.strictEqual(eng.generatedGarbage(3, false), 2);
    assert.strictEqual(eng.generatedGarbage(4, false), 4);
    assert.strictEqual(eng.generatedGarbage(4, true), 14); // + special (10)
  });

  test('cancellation arithmetic: [genGarbage, remGarbage], plus the caller-side remGenGarbage formula', () => {
    const eng = makeEngine();
    // clearedLines=2 -> gen=1 (2-1, c<4 branch); garbage=5 already covers it: rem=4, remGen=0.
    assert.deepStrictEqual(eng.genRemGarbage(5, 2, false), [1, 4]);
    assert.strictEqual(Math.max(0, 1 - 5), 0);
    // clearedLines=6 -> gen=6 (c>=4 branch); garbage=1 falls short: rem=0, remGen=5.
    assert.deepStrictEqual(eng.genRemGarbage(1, 6, false), [6, 0]);
    assert.strictEqual(Math.max(0, 6 - 1), 5);
  });
});

describe('fixPiece — materialization (§4-T7c)', () => {
  test('remGarbage >= HM: gameover via overflow, no negative-index scan, mg stays exactly HM rows', () => {
    const eng = makeEngine();
    const m = new eng.Machine(0, bagsFnFrom(['A', 'B']));
    m.garbage = TI.InitialMainGrid.length + 10; // force remGarbage > HM after a zero-clear fix
    const holesFn = makeHolesFn(TI.InitialMainGrid[0].length);
    // dropPiece relocates to the floor then fixes — already independently
    // tested (§ "dropPiece — calls T7's own fixPiece") — using it here avoids
    // re-deriving that relocation by hand a second time, and avoids
    // doesNotThrow silently passing over a false return (see below).
    let fired;
    assert.doesNotThrow(() => { fired = m.dropPiece(['A', 'B'], holesFn); });
    assert.strictEqual(fired, true, 'dropPiece must actually fire here — a resting piece must be fixable');
    assert.strictEqual(m.mg.length, TI.InitialMainGrid.length, 'mg must never grow past HM');
    assert.strictEqual(m.gameover, true);
  });

  test('holesFn is queried only for rows actually materialized (effRem), never beyond it', () => {
    const eng = makeEngine();
    const m = new eng.Machine(0, bagsFnFrom(['A', 'B']));
    m.garbage = 1;
    m.s6.s1.py = m.s6.gy; // same requirement — otherwise fixPiece is a guaranteed no-op
                          // and this test would pass vacuously (countingHolesFn never called)
    let calls = 0;
    const countingHolesFn = (y) => { calls++; return 0; };
    const fired = m.fixPiece(['A', 'B'], countingHolesFn);
    assert.strictEqual(fired, true, 'fixPiece must actually fire for this test to check anything');
    assert.ok(calls <= TI.InitialMainGrid.length);
  });
});

describe('Target round-robin — the 3-player worked trace from T7.v\'s own comment', () => {
  test('A -> B -> C initially; A -> C once B is gameover; A -> A (self) once C is also gameover', () => {
    const [A, B, C] = [0, 1, 2];
    const eng = makeEngine();
    const m = new eng.Machine(A, bagsFnFrom(['A', 'B']));
    assert.strictEqual(m.target, B, 'initial target is PlayerNext(self)');

    m.gameoverView[B] = true; // B observed gameover
    m.connectedView[B] = true;
    const viewAfterB = (pl2) => eng.playingView(m.gameoverView, m.connectedView, pl2);
    m.target = eng.nextTarget(viewAfterB, A, B); // mirrors receiveGameover's redirect from B
    assert.strictEqual(m.target, C);

    m.gameoverView[C] = true;
    const viewAfterC = (pl2) => eng.playingView(m.gameoverView, m.connectedView, pl2);
    m.target = eng.nextTarget(viewAfterC, A, C);
    assert.strictEqual(m.target, A, 'no one else alive: target becomes self (req-multi-target-nonself\'s winner exception)');
  });
});

describe('receiveGarbage / receiveGameover / receiveDisconnect — field updates', () => {
  test('receiveGarbage: sender-agnostic += , no from parameter', () => {
    const eng = makeEngine();
    const m = new eng.Machine(0, bagsFnFrom(['A', 'B']));
    m.receiveGarbage(3);
    m.receiveGarbage(2);
    assert.strictEqual(m.garbage, 5);
  });

  test('receiveGameover: sets the view, redirects target only if it pointed at the dead player', () => {
    const eng = makeEngine();
    const m = new eng.Machine(0, bagsFnFrom(['A', 'B']));
    assert.strictEqual(m.target, 1);
    m.receiveGameover(2); // target is 1, not 2 — no redirect
    assert.strictEqual(m.gameoverView[2], true);
    assert.strictEqual(m.target, 1);
    m.receiveGameover(1); // target is 1 — redirect
    assert.strictEqual(m.gameoverView[1], true);
    assert.notStrictEqual(m.target, 1);
  });

  test('receiveDisconnect: same redirect discipline as receiveGameover', () => {
    const eng = makeEngine();
    const m = new eng.Machine(0, bagsFnFrom(['A', 'B']));
    m.receiveDisconnect(1); // target is 1 — redirect
    assert.strictEqual(m.connectedView[1], false);
    assert.notStrictEqual(m.target, 1);
  });

  test('receiveX methods are not gated on winner/gameover — a disconnected/gameover player keeps draining', () => {
    const eng = makeEngine();
    const m = new eng.Machine(0, bagsFnFrom(['A', 'B']));
    m.s6.s1.gameover = true; // simulate self already gameover
    assert.doesNotThrow(() => { m.receiveGarbage(1); m.receiveGameover(1); m.receiveDisconnect(2); });
  });
});

describe('noticeDisconnection', () => {
  test('flips only the self-diagonal of connectedView', () => {
    const eng = makeEngine();
    const m = new eng.Machine(0, bagsFnFrom(['A', 'B']));
    m.noticeDisconnection();
    assert.strictEqual(m.connectedView[0], false);
    assert.strictEqual(m.connectedView[1], true);
    assert.strictEqual(m.connectedView[2], true);
  });
});

describe('winnerMulti gates every guarded method', () => {
  test('once winnerMulti is true, move/rotate/hold/kick/fix/fall/drop all return false with zero mutation', () => {
    const eng = makeEngine();
    const m = new eng.Machine(0, bagsFnFrom(['A', 'B']));
    m.gameoverView[1] = true;
    m.gameoverView[2] = true; // everyone else out — self is the winner
    assert.strictEqual(m.winner, true);
    const before = eng.snapshot(m);
    const holesFn = makeHolesFn(TI.InitialMainGrid[0].length);
    assert.strictEqual(m.movePiece(0, 1), false);
    assert.strictEqual(m.rotatePiece(true), false);
    assert.strictEqual(m.holdPiece(['A', 'B']), false);
    assert.strictEqual(m.rotateKickPiece(true), false);
    assert.strictEqual(m.fixPiece(['A', 'B'], holesFn), false);
    assert.strictEqual(m.fallStep(['A', 'B'], holesFn), false);
    assert.strictEqual(m.dropPiece(['A', 'B'], holesFn), false);
    assertSnapshotsEqual(eng.snapshot(m), before);
  });
});

describe('dropPiece — calls T7\'s own fixPiece, not this.s6.dropPiece (§4-T7i)', () => {
  test('a drop that generates garbage sends it (materialization not bypassed)', () => {
    const eng = makeEngine();
    const m = new eng.Machine(0, bagsFnFrom(['A', 'B']));
    m.garbage = 5; // pending garbage that a real fix should try to cancel
    const holesFn = makeHolesFn(TI.InitialMainGrid[0].length);
    const fired = m.dropPiece(['A', 'B'], holesFn);
    if (fired) {
      // whatever happened, garbage must have been processed through T7's own
      // accounting (reset to 0, or consumed into remGenGarbage) — never left
      // untouched at 5, which is what calling this.s6.dropPiece would do.
      assert.notStrictEqual(m.garbage, 5);
    }
  });
});

File: rs/t7/tests/model_unit_test.rs
// spec: tests/model_unit_test.rs — golden vectors (§9, `implementation.md`).
//
// Covers only what's new to T7; `t6`'s own single-player mechanics are
// exercised by `t6/tests/model_unit_test.rs`.
//
// Geometry (`TestInstance`: `PW=3`, `HM=6`, `WM=5`, `FY=4`, `FX=0`,
// `INITIAL_Y=3`, `INITIAL_X=1`): `mg[y][x]`, `y` increasing upward (`y=0`
// the floor). `Bar`'s `r0` occupies local row `dy=1` only, so on an empty
// board it drops all the way to absolute row `0` (`gy = -1`), landing at
// columns `px..px+2`.

#[path = "test_instance.rs"]
mod fixture;
#[path = "oracle.rs"]
mod oracle;

use fixture::{Piece, TestInstance, PLAYER_COUNT};
use t1::model::PieceOrExtra;
use t7::model::{
    check_invariants, gen_rem_garbage, generated_garbage, next_target, Garbage, Machine,
};

fn bags_alternating() -> impl FnMut(u64) -> Vec<Piece> {
    |i: u64| {
        if i.is_multiple_of(2) {
            vec![Piece::Bar, Piece::Corner]
        } else {
            vec![Piece::Corner, Piece::Bar]
        }
    }
}

fn piece_source() -> Piece {
    Piece::Bar
}

fn fresh(my_index: usize, player_count: usize) -> Machine<TestInstance> {
    Machine::<TestInstance>::new(my_index, player_count, bags_alternating(), piece_source)
}

fn no_holes() -> impl Fn(i64) -> i64 {
    |_y| 2 // constant hole column, well inside WM=5
}

// ── generated_garbage / gen_rem_garbage: golden truth table ─────────────

#[test]
fn generated_garbage_truth_table() {
    // spec: GeneratedGarbage — req-multi-garbage-gen
    assert_eq!(generated_garbage(0, false), 0); // 0 < 4: 0-1 -> max(0,-1) = 0
    assert_eq!(generated_garbage(1, false), 0); // 1-1 = 0
    assert_eq!(generated_garbage(2, false), 1);
    assert_eq!(generated_garbage(3, false), 2);
    assert_eq!(generated_garbage(4, false), 4); // >= 4: no -1
    assert_eq!(generated_garbage(5, false), 5);
    assert_eq!(generated_garbage(0, true), 10); // perfect clear special garbage
    assert_eq!(generated_garbage(4, true), 14);
}

#[test]
fn gen_rem_garbage_truth_table() {
    // spec: GenRemGarbage — (genGarbage, remGarbage), req-multi-garbage-cancel
    assert_eq!(gen_rem_garbage(0, 0, false), (0, 0));
    assert_eq!(gen_rem_garbage(5, 0, false), (0, 5)); // nothing generated, all of garbage remains
    assert_eq!(gen_rem_garbage(1, 2, false), (1, 0)); // generated (1) exactly cancels pending (1)
    assert_eq!(gen_rem_garbage(0, 2, false), (1, 0)); // nothing pending: remGarbage floors at 0
    assert_eq!(gen_rem_garbage(3, 2, false), (1, 2)); // partial cancellation
    assert_eq!(gen_rem_garbage(0, 0, true), (10, 0));
}

// ── materialization: unobstructed ────────────────────────────────────────

/// `drop_piece` with pending garbage below `HM`, on an otherwise-empty
/// board (no line clear, so `genGarbage = 0` and `remGarbage = garbage`
/// exactly): the fixed piece's own row is shifted up by exactly
/// `remGarbage`, and the bottom `remGarbage` rows become garbage rows with
/// the hole at the configured column.
#[test]
fn materialize_unobstructed_shifts_and_fills_garbage_rows() {
    let mut m = fresh(0, PLAYER_COUNT);
    // Force the piece to Bar r0 (single occupied row, dy=1) so the fixed
    // footprint is exactly row 0, cols 1..3, regardless of bag draw order.
    // py/px already start at INITIAL_Y/INITIAL_X; update_shadow_y refreshes
    // gy for Bar's shape.
    m.s6.s4.s3.s2.s1.p = Piece::Bar;
    m.s6.s4.s3.s2.s1.pr = 0;
    m.s6.update_shadow_y();
    m.garbage = 2;
    let bag = vec![Piece::Bar, Piece::Corner];

    let fired = m.drop_piece(&bag, no_holes());

    assert!(fired);
    let mg = &m.s6.s4.s3.s2.s1.mg;
    // row 2 = the fixed piece's own row (originally row 0), shifted up by 2
    assert_eq!(mg[2][0], None);
    assert_eq!(mg[2][1], Some(PieceOrExtra::Piece(Piece::Bar)));
    assert_eq!(mg[2][2], Some(PieceOrExtra::Piece(Piece::Bar)));
    assert_eq!(mg[2][3], Some(PieceOrExtra::Piece(Piece::Bar)));
    assert_eq!(mg[2][4], None);
    // rows 0, 1 = garbage rows, hole at column 2
    for (y, row) in mg.iter().enumerate().take(2usize) {
        for (x, cell) in row.iter().enumerate().take(5usize) {
            if x == 2 {
                assert_eq!(*cell, None, "row {y} hole");
            } else {
                assert_eq!(
                    *cell,
                    Some(PieceOrExtra::Extra(Garbage)),
                    "row {y} col {x}"
                );
            }
        }
    }
    // rows 3, 4, 5 = untouched, still empty
    for (y, row) in mg.iter().enumerate().take(6usize).skip(3) {
        assert!(row.iter().all(Option::is_none), "row {y} should be empty");
    }
    assert!(!m.s6.s4.s3.s2.s1.gameover);
    assert_eq!(m.garbage, 0); // consumed by materialization
    assert_eq!(m.rem_gen_garbage, 0); // nothing generated (no line cleared)
    check_invariants(&m);
}

// ── materialization: overflow (remGarbage >= HM) ────────────────────────

/// `remGarbage >= HM`: the entire board is pushed off — every row becomes a
/// garbage row and gameover fires unconditionally, no scan needed.
#[test]
fn materialize_overflow_when_remaining_at_least_hm() {
    let mut m = fresh(0, PLAYER_COUNT);
    m.garbage = 8; // >= HM = 6
    let bag = vec![Piece::Bar, Piece::Corner];

    let fired = m.drop_piece(&bag, no_holes());

    assert!(fired);
    let mg = &m.s6.s4.s3.s2.s1.mg;
    for row in mg.iter().take(6usize) {
        for (x, cell) in row.iter().enumerate().take(5usize) {
            if x == 2 {
                assert_eq!(*cell, None);
            } else {
                assert_eq!(*cell, Some(PieceOrExtra::Extra(Garbage)));
            }
        }
    }
    assert!(m.s6.s4.s3.s2.s1.gameover);
    assert!(m.gameover_view[0]);
    check_invariants(&m);
}

// ── materialization: forbidden-zone recheck after a non-overflowing shift ──

/// A pre-existing occupied cell (row 3, outside the falling piece's own
/// columns) is shifted into the forbidden zone (`FY=4`, `FH=2`) by garbage
/// insufficient to overflow the top. Gameover must still fire, via the
/// trailing forbidden-zone recheck on the new grid (§4-T7c).
#[test]
fn materialize_forbidden_zone_recheck_without_overflow() {
    let mut m = fresh(0, PLAYER_COUNT);
    m.s6.s4.s3.s2.s1.mg[3][4] = Some(PieceOrExtra::Piece(Piece::Corner)); // column 4: outside the piece's own cols 1..3
    m.garbage = 2;
    let bag = vec![Piece::Bar, Piece::Corner];

    let fired = m.drop_piece(&bag, no_holes());

    assert!(fired);
    let mg = &m.s6.s4.s3.s2.s1.mg;
    assert_eq!(
        mg[5][4],
        Some(PieceOrExtra::Piece(Piece::Corner)),
        "row 3 shifted to row 5 (3 + 2)"
    );
    assert!(
        m.s6.s4.s3.s2.s1.gameover,
        "forbidden-zone recheck must catch the shifted cell"
    );
    assert!(m.gameover_view[0]);
    check_invariants(&m);
}

// ── drop_piece == an equivalent fall_step sequence ending in a fix ──────

/// `drop_piece` must reach the same materialized result as `move_piece(-1,
/// 0)` stepped to rest, then `fix_piece` directly (§9).
#[test]
fn drop_piece_matches_equivalent_fall_sequence() {
    let bag = vec![Piece::Bar, Piece::Corner];

    let mut dropped = fresh(0, PLAYER_COUNT);
    dropped.garbage = 3;
    assert!(dropped.drop_piece(&bag, no_holes()));

    let mut stepped = fresh(0, PLAYER_COUNT);
    stepped.garbage = 3;
    while stepped.move_piece(-1, 0) {}
    assert!(stepped.fix_piece(&bag, no_holes()));

    assert_eq!(dropped.s6.s4.s3.s2.s1.mg, stepped.s6.s4.s3.s2.s1.mg);
    assert_eq!(
        dropped.s6.s4.s3.s2.s1.gameover,
        stepped.s6.s4.s3.s2.s1.gameover
    );
    assert_eq!(dropped.garbage, stepped.garbage);
    assert_eq!(dropped.rem_gen_garbage, stepped.rem_gen_garbage);
    assert_eq!(dropped.target, stepped.target);
    check_invariants(&dropped);
    check_invariants(&stepped);
}

// ── target round-robin ───────────────────────────────────────────────────

/// `next_target` directly: a 3-player cycle, `self=0`, starting at `pl=0`,
/// skips a not-playing candidate and lands on the next one that is.
#[test]
fn next_target_skips_not_playing_and_self() {
    let playing = |pl2: usize| pl2 != 1; // player 1 not playing
    assert_eq!(next_target(playing, 0, 0, 3), 2); // 0 -> 1 (not playing) -> 2 (playing, != self)
}

/// `next_target` falls back to `self` when nobody else is playing (the
/// winner case — req-multi-target-nonself).
#[test]
fn next_target_falls_back_to_self_when_no_candidate() {
    let playing = |pl2: usize| pl2 == 0; // only self is playing
    assert_eq!(next_target(playing, 0, 0, 3), 0);
}

/// `receive_gameover` redirects `target` away from a now-dead player,
/// walking the 3-player cycle; once no other player remains, `target`
/// falls back to `self` (`T7.v`'s `NextTargetAux` — not a bug).
#[test]
fn receive_gameover_redirects_target_round_robin() {
    let mut m = fresh(0, 3);
    assert_eq!(m.target, 1); // Init: target = PlayerNext(myIndex) = 1

    m.receive_gameover(1);
    assert!(m.gameover_view[1]);
    assert_eq!(
        m.target, 2,
        "target redirects from the now-dead player 1 to player 2"
    );

    m.receive_gameover(2);
    assert!(m.gameover_view[2]);
    assert_eq!(
        m.target, 0,
        "no other player left playing -> falls back to self"
    );
}

// ── receive_* field updates ──────────────────────────────────────────────

#[test]
fn receive_garbage_accumulates() {
    let mut m = fresh(0, PLAYER_COUNT);
    assert_eq!(m.garbage, 0);
    m.receive_garbage(3);
    assert_eq!(m.garbage, 3);
    m.receive_garbage(4);
    assert_eq!(m.garbage, 7);
}

#[test]
fn receive_disconnect_updates_view_and_redirects_target() {
    let mut m = fresh(0, 3);
    assert_eq!(m.target, 1);
    m.receive_disconnect(1);
    assert!(!m.connected_view[1]);
    assert_eq!(
        m.target, 2,
        "target redirects away from the now-disconnected player 1"
    );
}

// ── notice_disconnection: idempotency ────────────────────────────────────

#[test]
fn notice_disconnection_idempotent() {
    let mut m = fresh(0, PLAYER_COUNT);
    assert!(m.connected_view[0]);
    assert!(m.notice_disconnection(), "first call fires");
    assert!(!m.connected_view[0]);
    assert!(
        !m.notice_disconnection(),
        "second call is a no-op — already noticed"
    );
    assert!(!m.connected_view[0]);
}

// ── §4-T7h: drop_piece's winner_multi guard leaves py/px untouched ──────

/// A `WinnerMulti` `drop_piece` must leave `py`/`px` byte-identical, even
/// transiently (§4-T7h).
#[test]
fn drop_piece_winner_multi_guard_leaves_state_untouched() {
    let mut m = fresh(0, 2);
    m.gameover_view[1] = true; // the only other player is gameover -> self is the winner
    assert!(t7::model::winner_multi(
        &m.gameover_view,
        &m.connected_view,
        m.my_index
    ));

    let before = (
        m.s6.s4.s3.s2.s1.py,
        m.s6.s4.s3.s2.s1.px,
        m.s6.s4.s3.s2.s1.mg.clone(),
    );
    let bag = vec![Piece::Bar, Piece::Corner];

    let fired = m.drop_piece(&bag, no_holes());

    assert!(!fired);
    assert_eq!(m.s6.s4.s3.s2.s1.py, before.0);
    assert_eq!(m.s6.s4.s3.s2.s1.px, before.1);
    assert_eq!(m.s6.s4.s3.s2.s1.mg, before.2);
}

// ── differential: one hand-traced materialization case agrees with oracle.rs ──

#[test]
fn materialize_case_matches_independent_oracle() {
    let mut m = fresh(0, PLAYER_COUNT);
    m.s6.s4.s3.s2.s1.p = Piece::Bar; // force a known, single-row footprint — see the test above
    m.s6.s4.s3.s2.s1.pr = 0;
    m.s6.update_shadow_y();
    m.s6.s4.s3.s2.s1.mg[3][4] = Some(PieceOrExtra::Piece(Piece::Corner));
    let old_mg = m.s6.s4.s3.s2.s1.mg.clone();
    m.garbage = 2;
    let bag = vec![Piece::Bar, Piece::Corner];

    assert!(m.drop_piece(&bag, no_holes()));

    // The oracle only models materialization itself, so replay it against
    // the pre-fix grid plus the piece's fixed row (row 0, cols 1..3) to
    // reconstruct `self.s6.fix_piece`'s output just before materialization.
    let mut pre_materialize = old_mg;
    pre_materialize[0][1] = Some(PieceOrExtra::Piece(Piece::Bar));
    pre_materialize[0][2] = Some(PieceOrExtra::Piece(Piece::Bar));
    pre_materialize[0][3] = Some(PieceOrExtra::Piece(Piece::Bar));
    let (expected_mg, expected_overflow) = oracle::oracle_materialize(
        &pre_materialize,
        2,
        no_holes(),
        PieceOrExtra::Extra(Garbage),
    );

    assert_eq!(m.s6.s4.s3.s2.s1.mg, expected_mg);
    assert!(!expected_overflow);
    assert!(
        m.s6.s4.s3.s2.s1.gameover,
        "forbidden-zone recheck still catches the shifted cell"
    );
}
